jobs_archival/worker.go

```diff
@@ -0,0 +1,228 @@
+		return false
+	}
+	limitReached = failed.LimitsReached
+	jobs = failed.Jobs
+	if !limitReached {
+		params.EventsLimit -= failed.EventsCount
+		params.PayloadSizeLimit -= failed.PayloadSize
+		unProcessed, err = w.jobsDB.GetUnprocessed(w.lifecycle.ctx, params)
+		if err != nil {
+			w.log.Errorf("failed to fetch unprocessed jobs for backup - partition: %s - %w", w.partition, err)
+			return false
+		}
+		jobs = append(jobs, unProcessed.Jobs...)
+		limitReached = unProcessed.LimitsReached
+	}
+
+	if len(jobs) == 0 {
+		return false
+	}
+
+	storagePrefs, err := w.storageProvider.GetStoragePreferences(jobs[0].WorkspaceId)
+	var reason string
+	if err != nil {
+		w.log.Errorf("failed to fetch storage preferences for workspaceID: %s - %w", jobs[0].WorkspaceId, err)
+		reason = fmt.Sprintf(`{"location": "not uploaded because - %v"}`, err)
+		toArchive = false
+	}
+	if !storagePrefs.Backup(w.archiveFrom) {
+		reason = fmt.Sprintf(`{"location": "not uploaded because storage disabled for %s"}`, w.archiveFrom)
+		toArchive = false
+	}
+	var statusList []*jobsdb.JobStatusT
+	if !toArchive {
+		statusList = getStatuses(
+			jobs,
+			func(*jobsdb.JobT) string { return jobsdb.Aborted.State },
+			[]byte(reason),
+		)
+		goto markStatus
+	}
+
+	res = w.uploadJobs(w.lifecycle.ctx, jobs)
+	if res.err != nil {
+		w.log.Errorf("failed to upload jobs - partition: %s - %w", w.partition, res.err)
+		maxAttempts := maxRetryAttempts()
+		statusList = getStatuses(
+			jobs,
+			func(job *jobsdb.JobT) string {
+				if job.LastJobStatus.AttemptNum >= maxAttempts {
+					return jobsdb.Aborted.State
+				}
+				return jobsdb.Failed.State
+			},
+			[]byte(fmt.Sprintf(`{"location": "not uploaded because - %v"}`, res.err)),
+		)
+		goto markStatus
+	}
+
+	statusList = getStatuses(
+		jobs,
+		func(*jobsdb.JobT) string { return jobsdb.Succeeded.State },
+		[]byte(fmt.Sprintf(`{"location": "%v"}`, res.location)),
+	)
+
+markStatus:
+	if err := w.jobsDB.UpdateJobStatus(w.lifecycle.ctx, statusList, nil, nil); err != nil {
+		w.log.Errorf("failed to mark jobs' status - %w", err)
+	}
+	if limitReached {
+		goto start
+	}
+	return true
+}
+
+func (w *worker) SleepDurations() (min, max time.Duration) {
+	return 0, 0
+}
+
+func (w *worker) Stop() {
+	w.lifecycle.cancel()
+}
+
+type uploadResult struct {
+	location string
+	err      error
+}
+
+func (w *worker) uploadJobs(ctx context.Context, jobs []*jobsdb.JobT) uploadResult {
+	firstJobCreatedAt := jobs[0].CreatedAt
+	lastJobCreatedAt := jobs[len(jobs)-1].CreatedAt
+	workspaceID := jobs[0].WorkspaceId
+
+	w.log.Infof("[Archival: storeErrorsToObjectStorage]: Starting logging to object storage - %s", w.partition)
+
+	gzWriter := fileuploader.NewGzMultiFileWriter()
+	path := fmt.Sprintf(
+		"%v%v.json.gz",
+		lo.Must(misc.CreateTMPDIR())+"/rudder-backups/"+w.partition+"/",
+		fmt.Sprintf("%v_%v_%v", firstJobCreatedAt.Unix(), lastJobCreatedAt.Unix(), workspaceID),
+	)
+
+	for _, job := range jobs {
+		rawJob, err := json.Marshal(job)
```

Comment:
true
we may only need the event payload in that case.
The others are not used at sources imo, correct me if I'm wrong.

Reply:
We do need to find a way to distribute load across rudder nodes during event replays... 
Instead of relying on the `UserID` for that we could simply use the node's identifier and batch messages more efficiently, but including the `UserID` column wouldn't hurt us, in case we ultimately need it. `CreatedAt` might be useful too, e.g. as a fallback to `receivedAt`.